algopy_testing/src/algopy/primitives/biguint.py

```diff
@@ -0,0 +1,163 @@
+from __future__ import annotations
+
+import functools
+
+from algopy.constants import MAX_UINT512, UNIT64_BYTES_LENGTH
+from algopy.primitives.bytes import Bytes
+from algopy.primitives.uint64 import UInt64
+
+# TypeError, ValueError are used for operations that are compile time errors
+# ArithmeticError and subclasses are used for operations that would fail during AVM execution
+
+
+@functools.total_ordering
+class BigUInt:
+    """
+    A python implementation of an TEAL bigint type represented by AVM []byte type
+    """
+
+    __value: bytes  # underlying 'bytes' value representing the BigUInt
+
+    def __init__(self, value: UInt64 | int = 0) -> None:
+        self.__value = (
+            _int_to_bytes(value.value, UNIT64_BYTES_LENGTH)
+            if isinstance(value, UInt64)
+            else _int_to_bytes(value)
+        )
+
+    def __repr__(self) -> str:
+        return f"{self.value}u"
+
+    def __str__(self) -> str:
+        return str(self.value)
+
+    # comparison operations should not error when comparing any uint type
+    # e.g. UInt64, BigUInt, arc4.UInt*, arc4.BigUint*
+    # however will raise a TypeError if compared to something that is not a numeric value
+    # as this would be a compile error when compiled to TEAL
+    def __eq__(self, other: object) -> bool:
+        return _as_int512(self) == _as_int512(other)
+
+    def __lt__(self, other: BigUInt | UInt64 | int) -> bool:
+        return _as_int512(self) < _as_int512(other)
+
+    def __bool__(self) -> bool:
+        return bool(_as_int512(self))
+
+    def __add__(self, other: BigUInt | UInt64 | int) -> BigUInt:
+        return _checked_result(_as_int512(self) + _as_int512(other), "+")
```

Comment:
In order to match AVM behaviour, I think we need to remove the `_checked_result` call here so that an overflow in the result does not cause an error.

Instead an error should occur if the inputs to an arithmetic function have been exceeded i.e. if `_as_int512` returns a value > MAX_UINT512

Reply:
Thanks Dan. It has been fixed in [this commit](https://github.com/algorandfoundation/puya/pull/218/commits/51d0a443d27076504a0b1eb3c2fb62a54696311c).